tests/test_snapcraft_store_api.py

```diff
@@ -0,0 +1,230 @@
+from urllib.parse import urlencode
+from vcr_unittest import VCRTestCase
+from canonicalwebteam.store_api.stores.snapcraft import SnapcraftStoreApi
+
+
+class SnapcraftApiTest(VCRTestCase):
+    def setUp(self):
+        # configure self.attribute
+        self.client = SnapcraftStoreApi()
+        return super(SnapcraftApiTest, self).setUp()
+
+    def test_get_all_items(self):
```

Comment:
While it's laudable that you've got all these assertions here I feel that they may simply be creating tests that are less resilient and harder to update.

What you really want to test is that:
- Given a specific api call `self.client.get_all_items(size=16, api_version=1)`
- You are returned the correct response

The url, headers, and response code are all superfluous to this goal. If any changes are made to the api call signatures then you will have to update all these tests instead of simply re-generating the `vcr`. You shouldn't need to care about how the function operates, just that it provides the correct response.

Reply:
Yes I'm afraid I agree with @hatched, and so what I'm going to write might just be reiterating what he's said but in my own language. Sorry if that's the case, but I think this is important so I'm going to write it anyway in the hope it can be educational.

What you're doing in the below code is calling a function (`get_all_items`) and then testing a few *internal specifics* and quirks of what you know that function should have done - requested a specific exact URL, got a response from the API containing a `X-Ubuntu-Series` header. I hope this doesn't come across as too mean, but this is kind of the exact opposite of what a good test should be.

It's basically a rule of testing that you don't test internals, and I personally think this is a really central principle of testing. You want your test to make sure your module's *public interface* is providing *the functionality that it promises to provide* (this basically builds on the OOP concept of [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming))).

To take your `get_all_items` as an example, this means that you want to check that the function succeeds in getting all items, and that it responds appropriately to any arguments you pass it - if you change "size", it limits the number of items you get back, etc. (I assume that's what that does). Conversely, what you don't want to check at all is stuff that is invisible to the user of the module. This includes the URL that was requested, the format of the API response, the API response status code - these are all details that are internal to the `get_all_items` function. They could completely change and it wouldn't matter, as long as the function still performed its core function, which is to get all items.

The reason this matters so much is that if you do a lot of testing internals - if the tests are relying on the exact way functions are implemented - you make the code practically impossible to change without breaking tests all over the place. This can lead to developers being scared to make significant changes in a project's code, which can be the death of a project as it just becomes a spaghetti pile of monkey patch in top of monkey patch.

So we should always strive to test the important things as effectively as possible while leaving the maximum flexibility for the code to change internally without effecting the tests. So we test the public interface.

Reply:
Hey, I made changes into the tests. I think you are right about this. If you don't mind it would be nice to have your input again. Thank you for your time.

Reply:
Yep this is much more what I was thinking of